server/src/daemon/service.rs

```diff
@@ -0,0 +1,368 @@
+        let secret = Secret::from(password.into_bytes());
+        Self {
+            collections: Arc::new(RwLock::new(Vec::new())),
+            keyring: Arc::new(Keyring::load(path, secret).await.unwrap()),
+            cnx: Default::default(),
+            manager: Arc::new(Mutex::new(ServiceManager::default())),
+            sessions_counter: Arc::new(RwLock::new(0)),
+        }
+    }
+
+    pub async fn run(self) -> Result<()> {
+        #[cfg(debug_assertions)]
+        let service_name = SERVICE_NAME;
+        #[cfg(not(debug_assertions))]
+        let service_name = oo7::dbus::api::Service::DESTINATION.unwrap();
+        let cnx = zbus::connection::Builder::session()?
+            .name(service_name)?
+            .build()
+            .await?;
+        *self.cnx.lock().unwrap() = Some(cnx.clone());
+        let object_server = cnx.object_server();
+        object_server
+            .at(oo7::dbus::api::Service::PATH.unwrap(), self.clone())
+            .await?;
+
+        // loading login.keyring into the objects tree
+        Service::init_login(&object_server, self).await;
+
+        Ok(())
+    }
+
+    pub async fn init_login(object_server: &ObjectServer, service: Service) {
+        let path = format!("{}/{}", env::var("HOME").unwrap(), LOGIN_KEYRING_PATH);
+        let created = fs::metadata(path)
+            .unwrap()
+            .created()
+            .unwrap()
+            .duration_since(SystemTime::UNIX_EPOCH)
+            .unwrap();
+        let login = Collection::new(
+            LOGIN_KEYRING,
+            "login",
+            created,
+            service.keyring.clone(),
+            service.manager.clone(),
+        );
+        service.collections.write().await.push(login.clone());
+        let path = OwnedObjectPath::from(login.path());
+        object_server.at(&path, login.clone()).await.unwrap();
+
+        let n_items = service.keyring.n_items().await;
+        if n_items > 0 {
+            let mut items: Vec<Item> = Vec::with_capacity(n_items);
+            for item in service.keyring.items().await {
+                items.push(match item {
+                    Ok(item) => item,
+                    Err(e) => panic!("Item cannot be decrypted: {}", e.to_string()),
+                })
+            }
+            for item in items {
+                let item = item::Item::new(
+                    item.clone(),
+                    item.secret().to_vec(),
+                    "text/plain".to_string(),
+                    login.item_counter().await,
+                    login.path(),
+                    service.keyring.clone(),
+                    service.manager.clone(),
+                )
+                .await;
+                login.set_item_counter().await;
+                let path = OwnedObjectPath::from(item.path());
+                login.items.write().await.push(item.clone());
+                object_server.at(&path, item).await.unwrap();
```

Comment:
the item should have a method to export itself on an ObjectServer, this way you avoid exposing the path itself